### PR TITLE
Proposal: Document defining Gutenberg Principles

### PR DESCRIPTION
**Pull request by nerrad**
I've been involved with the Gutenberg project for a few years in different contexts: contributing directly to the codebase, participating in conversations around various interfaces, consuming what others have written about the project, writing plugins and blocks extending Gutenberg, consulting with teams working on it, and teaching others various aspects of the project. While there are various approaches to Gutenberg that are indirectly "picked up" by those working on it extensively as a by product of being "in the trenches", I have found that there's no real authoritative source for describing the underlying principles guiding Gutenberg development and design.

This lack manifests itself with some difficulty when trying to:

- communicate why certain architectural decisions are made in interfaces built.
- understand from a business perspective why Gutenberg is beneficial and why its approach should be considered in building for the WordPress ecosystem.
- help third parties who will be building extensions on top of the various GB interfaces we build and why things work “differently” than what they are used to in the WordPress ecosystem.

While there are a number of posts spread through various blogs and sources that articulate pieces of “Gutenberg principles”, to date there really isn’t a cohesive document than serves as an “official” repository of this guiding knowledge.

Having official principles guiding Gutenberg development can help in the following ways:

- Reviewers have something to help use when assessing the work in submitted PRs in the Gutenberg project.
- Guiding architectural discussions and decisions around changes or improvements in general design (both user and code). Besides having a clearer idea around the “yes” things we do, it also makes it much easier to identify why we might say “no” to an idea or approach.
- Communicating with the wider development and WordPress business community that might have competing interests for how they feel Gutenberg “should be done”. This provides something to align their own work with.

I believe there is an underlying set of ideas and principles that are guiding Gutenberg development and design whether it's communicated officially or not.

Given the above, and my own understanding of the Gutenberg project from my participation, observation, and usage, I'm submitting a proposed document with what I think some of the principles are guiding GB development and design.  Some additional considerations:

- I'd love to have more variety and diversity represented in resources that complement the principles outlined in the doc before publishing. So please pop in some links for consideration if anyone knows of any.
- Related to the above, I think it'd be beneficial to consider this as an evolving doc. Even after publishing, we may find that certain principles can be refined or additional ones pop up. I also think it's beneficial to have this as an _entry point_ type of document that inspires folks to dive deeper into the implications of these principles in code, design or other posts published in our ecosystem. Ideally we could link to these various voices in the WordPress ecosystem from this doc as well over time.
- More than anything, this is intended to kickstart discussion around having these principles articulated and published. Whether or not the final form of this document ends up being what I've introduced here is less important to me than having _something_ in place.

### CODE REVIEW

File: docs/explanations/principles.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+
+## Aim for delightful, consistent, accessible user interfaces for empowering content creation and expression.
+
+_“Optimize for the user” (Matias - [Gutenberg or the Ship of Theseus](https://matiasventura.com/post/gutenberg-or-the-ship-of-theseus/))_
+
+Each of the words in this principle are deliberate:
+
+
+### Delightful
```
Comment (draganescu) [inline docs/explanations/principles.md]: As much as I love this word, it's just not a good principle because it is so subjective. It has also been confiscated by commercial projects.

Comment (nerrad) [reply docs/explanations/principles.md]: Is it primarily the specific word you object to or also the meaning behind the choice of the word (and the intent of the principle)? If just the word itself, do you have any suggestions about what could be used instead?

**Comment by youknowriad** [comment]
This is a great document. I don't have much to add to be honest :)

File: docs/explanations/principles.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+**Modularity** - the concept of being able to combine blocks or components into different types of layout or content (see “[Embrace the modularity](https://riad.blog/2020/01/28/embrace-the-modularity/)”).
+
+**Encapsulation** - there are appropriate error handling and boundaries around things that could be broken by things extending the interface.
+
+**Experimentation** - Often new APIs and interfaces clearly start as experimental while reliability and usability of different approaches are being tested. These experiments always begin with determining what problem is solved from the _creator_ perspective and seek to validate how it impacts the user workflow. The intent is to explore what generalizations of the creator's use-case are possible and to discover better abstractions before solidifying around the contract exposed to extension developers.
+
+
+## Progressive Complexity
+
+_“On the Layers of the Onion”_ (Matias - Gutenberg or the Ship of Theseus)
```
Comment (gziolo) [inline docs/explanations/principles.md]: It isn't linked.

File: docs/explanations/principles.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+## Ubiquitous and Safe Adoption
+
+This encompasses the idea of _backwards compatibility_ and _graceful degradation_.
+
+“One’s existing content should never be lost or generally affected by the switch to a new editor. The newer tool should understand the content natively or at the very least leave it undisturbed.” (Backwards compatibility principle in “[The Language of Gutenberg](https://lamda.blog/2018/04/22/the-language-of-gutenberg/)” - Miguel)
+
+This also encompasses the ideas of _portability and freedom from lock-in_:
+
+“One’s content shouldn’t be tied to anything. This means that content shouldn’t be tied to any runtime, be it Gutenberg or WordPress in general, and that a good relationship with other editors (mobile apps, MarsEdit, etc.) should be developed.” (Portability principle in “The Language of Gutenberg - Miguel)
```
Comment (gziolo) [inline docs/explanations/principles.md:76]: Link is missing.

File: docs/explanations/principles.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+“Adopting Gutenberg for early testing or curiosity shouldn’t be an irreversible action for one’s content. Disabling Gutenberg shouldn’t result in unreasonably altered content. The same should later apply to third-party blocks. There is a strong desire to fight lock-in, which is not new but is pervasive in a world of proprietary software and software-as-a-service.” (No commitment principle in “The language of Gutenberg - Miguel)
```
Comment (gziolo) [inline docs/explanations/principles.md:78]: Link is missing.

File: docs/explanations/principles.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+This also embraces the promise of the Web where content is replicable in multiple ways across devices and apps. It’s the idea that the basic block structure proves to be a timeless and reliable way of representing the content and meta information about that content.
+
+Imagine a world where content created by Gutenberg can be consumed and read _anywhere_. This is largely why at the heart of the block structure, the format is largely an enhancement on top of HTML.
+
+“Ultimately, choosing HTML means that — as with a painting or a sculpture — the editor’s final artefact is the canonical format of the content, not a byproduct thereof.” (The Language of Gutenberg - Miguel).
```
Comment (gziolo) [inline docs/explanations/principles.md]: Link is missing.

File: docs/explanations/principles.md [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+This manifests in letting machines do what machines are good for, and preserve content in a format that is readable to users.
+
+Finally, the principle of ubiquitous and safe adoption is found in the way Gutenberg is being developed _incrementally_. WordPress powers ~ 40% of the Web and the changes that GB brings cannot happen overnight (Gutenberg or the ship of Theseus - Matías)
```
Comment (gziolo) [inline docs/explanations/principles.md]: Link is missing.

And maybe we use "over 40%" so we don't have to change it too often :trollface: 

```suggestion
Finally, the principle of ubiquitous and safe adoption is found in the way Gutenberg is being developed _incrementally_. WordPress powers over 40% of the Web and the changes that GB brings cannot happen overnight (Gutenberg or the ship of Theseus - Matías)
```

**Review by gziolo — APPROVED** [review]
This is a great document. @nerrad, you did an extraordinary job to put it all together in such a thoughtful way.

I left some minor thing like missing links that I could spot when browsing:

https://github.com/WordPress/gutenberg/blob/add/gb-principles-doc/docs/explanations/principles.md

It would be great to have a few more ✅ to ensure we all share the same principles 😄

File: docs/explanations/principles.md (hunk 1 above)
Comment (tellthemachines) [reply docs/explanations/principles.md]: Delightful does sound a bit like marketing-speak. It's not so much the subjectivity as the underlying idea that we're somehow directing the user's feelings about the product that makes me uneasy.

The word "transparent" popped into my head when reading `The user interfaces should “get out of the way” and support the user`. Transparency both in the sense of clarity (it's clear how to use the interface, thus "supporting" the user instead of making things difficult) but also in the sense of invisibility (it gets "out of the way" so users don't even notice it while working).

Comment (nerrad) [reply docs/explanations/principles.md]: You both have shared perspectives on the word "delightful" that I didn't even think of. I was thinking of it in the context of aiming for user experiences that elicit "surprised joy" (my interpretation of "delight"). I also don't think we can escape the reality that we indeed _do_ influence users feelings about the product when building it.  This principle is aimed at thinking how what we build contributes to positive feelings.

With that said, I do see how **transparent** could be a good substitute here for the word delightful and moves more towards the objectiveness of a _principle_ as opposed to the subjectivity of a value or goal.

Comment (aristath) [reply docs/explanations/principles.md]: +1 for `transparent` :+1:

File: docs/README.md [hunk 7]
```diff
@@ -20,6 +20,9 @@ The Editor offers rich new value to users with visual, drag-and-drop creation to
 
 ## Quick links
 
+### Gutenberg Principles
+[The Gutenberg principles document](/docs/explanations/principles.md) highlights a few principles that are guiding the overall project and shape various development and design decisions.
+
```
Comment (mkaz) [inline docs/README.md]: Can we move this section into the `Contribute to the block editor` below, I don't think it should be the first Quick Link.

This is the first page that shows when viewing the [Block Editor Handbook](https://developer.wordpress.org/block-editor/) and people are going to be more interested in how to build for and extend first.

With that also said, I think a better more prominent spot would be the [repository top-level readme](https://github.com/WordPress/gutenberg/blob/trunk/README.md), as that is only showed within the repo and more likely target to developers of the project.

**Review by mkaz — COMMENTED** [review]
Since this is a new document you need to add it to `docs/toc.json` and then `npm run docs:build` and then commit the updated `manifest.json` file. :+1: 

See the existing listings in `toc.json` for syntax.

**Comment by mtias** [comment]
I'd like to give this a bit more thought before merging.

File: docs/explanations/principles.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+## Curated Extensibility
+
+The use of the word _curated_ expressly refers to implementing extensibility in a way that prioritizes those _using Gutenberg interfaces to create their content_ over developers integrating with the system. As a result, every extensibility interface request is filtered through the question of how it impacts creator workflows.
+
+Implementation details are ephemeral. Practically this means that the traditional WP approach of wrapping things in filters and liberal use of actions is avoided in favor of explicit extension interfaces that expose _functionality_ but hide the _implementation details._ Gutenberg has opinionated APIs that favor stability and functionality for the user over flexibility for the developer.
+
+Extensibility interfaces in Gutenberg favor:
+
+**Modularity** - the concept of being able to combine blocks or components into different types of layout or content (see “[Embrace the modularity](https://riad.blog/2020/01/28/embrace-the-modularity/)”).
```
Comment (mtias) [inline docs/explanations/principles.md:40]: There's some overlap here with https://developer.wordpress.org/block-editor/explanations/architecture/key-concepts/ — it should be linked at least

File: docs/README.md [hunk 9]
```diff
@@ -20,6 +20,9 @@ The Editor offers rich new value to users with visual, drag-and-drop creation to
 
 ## Quick links
 
+### Gutenberg Principles
```
Comment (mtias) [inline docs/README.md:23]: I think "Guiding Principles" would be better.

File: docs/explanations/principles.md [hunk 10]
```diff
@@ -0,0 +1,106 @@
+
+
+# Gutenberg Principles
```
Comment (mtias) [inline docs/explanations/principles.md]: I think "Guiding Principles" would be better.

File: docs/explanations/principles.md [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+This document has been written so that those contributing to the Gutenberg project can have some understanding of what guides decisions around what is included in the project (or not).
+
+An important thing to keep at top of mind when reviewing these principles is that they are not isolated from each other. There is a cohesiveness and complementary nature between them that results in a synergistic impact and direction.
+
+When applied, sometimes these principles build on each other much like the idea of constructing a house with a foundation, walls and roof. However, other times the application of the principles results in something that is more similar to a tapestry where threads of each principle are embedded together and evident in the finished feature or design of the project.
+
+
+## Aim for delightful, consistent, accessible user interfaces for empowering content creation and expression.
```
Comment (mtias) [inline docs/explanations/principles.md]: I adhere to "empowering content creation and expression" but I'm not sure the other concepts capture the right concepts. 🤔

File: docs/explanations/principles.md (hunk 1 above)
Comment (mtias) [reply docs/explanations/principles.md]: "Delightful" doesn't capture what I think this section hints at. "Transparent" is also lacking a bit in concrete meaning.

The way I'd describe this is "A canvas for expression" (more on that here: https://github.com/WordPress/gutenberg/issues/18667)

File: docs/explanations/principles.md [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+Each of the words in this principle are deliberate:
+
+
+### Delightful
+
+The user interfaces should “get out of the way” and support the user in creating their content. They should lead to experiences where the user (in the vernacular of Marie Kondo) _finds delight_ with each new thing they discover they can do or accomplish.
+
+
+### Consistent
```
Comment (mtias) [inline docs/explanations/principles.md]: This is not a principle I'd be comfortable stating as a primary guiding force because it's a bit more nuanced. The way I tend to describe this is more along the lines of "Providing uniformity within diversity, as learning the ingredients of the interface happens once but scales to hundreds of blocks."

This is an important design tension.

File: docs/explanations/principles.md [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+
+Any interfaces are consistently applied. Users should be able to _rely on muscle memory and learned patterns_ as they become familiar with how it works.
+
+This also applies to learning _where to find_ different elements that the user might interact with in creating their content.
+
+
+### Accessible
```
Comment (mtias) [inline docs/explanations/principles.md:26]: This one feels a bit odd as it's more of a general principle of WordPress at wide and not specific to Gutenberg. If anything, it should build upon the two other axis of "canvas for expression" and "uniform yet diverse" to embrace the fact users interacting with the software have distinct characteristics and preferences for how to use said software.

File: docs/explanations/principles.md [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+Extensibility interfaces in Gutenberg favor:
+
+**Modularity** - the concept of being able to combine blocks or components into different types of layout or content (see “[Embrace the modularity](https://riad.blog/2020/01/28/embrace-the-modularity/)”).
+
+**Encapsulation** - there are appropriate error handling and boundaries around things that could be broken by things extending the interface.
+
+**Experimentation** - Often new APIs and interfaces clearly start as experimental while reliability and usability of different approaches are being tested. These experiments always begin with determining what problem is solved from the _creator_ perspective and seek to validate how it impacts the user workflow. The intent is to explore what generalizations of the creator's use-case are possible and to discover better abstractions before solidifying around the contract exposed to extension developers.
+
+
+## Progressive Complexity
```
Comment (mtias) [inline docs/explanations/principles.md]: This needs to be balanced a bit more with a general seek for "simplicity"

File: docs/explanations/principles.md [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+
+An example here would be creators discovering they can move a block into another block in various contexts and discovering it _just works_. So even learning basic interactions can help _empower_ more complex interactions through the delight of experimenting as a natural outflow of the creative process. A great example of this is the power of multi-block transforms.
+
+There’s also the related application here to design systems in larger organizations where the design team wants to put some restrictions on what can be modified and manipulated by the content creation team in keeping with the chosen designs behind the presentation of that content. In this case, the progressive complexity principle in Gutenberg provides the tools for “locking down” various elements of the editor when and as needed depending on who is working with the content. Entire swaths of the interface could be hidden or removed to account for the fluent needs of those interacting with the content.
+
+
+## Ubiquitous and Safe Adoption
```
Comment (mtias) [inline docs/explanations/principles.md:68]: I this is more about a profound respect for the user in an expansive sense — both in how it respects where they come from (compatibility) and aims to ensure protections for the future (the way user data is treated, for example)

File: docs/explanations/principles.md (hunk 13 above)
Comment (mtias) [reply docs/explanations/principles.md:26]: There's also something important missing here which is how the software can help users build better websites that are semantic and more accessible. That is, _respect for the viewers as well as the creators_.

File: docs/explanations/principles.md (hunk 1 above)
Comment (nerrad) [reply docs/explanations/principles.md]: I'll see if I can capture some of what you wrote in that issue Matías - ❤️ it.

File: docs/explanations/principles.md (hunk 11 above)
Comment (nerrad) [reply docs/explanations/principles.md]: What about something like:

> The user experience is a canvas that empowers content creation and expression through intuitive and exploration friendly interfaces

File: docs/explanations/principles.md (hunk 12 above)
Comment (nerrad) [reply docs/explanations/principles.md]: I like the concept of "providing uniformity within diversity" but I think there's still the importance here of capturing the idea of not just the _block interfaces_ but also the interfaces within the editor? Would you say this statement is a principle that also applies to the editor interfaces?

Also, is this something that you believe should be explicitly called out as it's _own_ principle, or should it be included somehow as a part of [the statement commented on here](https://github.com/WordPress/gutenberg/pull/30963#discussion_r620302895)?

File: docs/explanations/principles.md (hunk 13 above)
Comment (nerrad) [reply docs/explanations/principles.md:26]: > That is, respect for the viewers as well as the creators.

This is definitely something that isn't capture well in the current iteration that should be included ❤️ 

> This one feels a bit odd as it's more of a general principle of WordPress at wide and not specific to Gutenberg. If anything, it should build upon the two other axis of "canvas for expression" and "uniform yet diverse" to embrace the fact users interacting with the software have distinct characteristics and preferences for how to use said software.

While I agree it's a general principle of WordPress, I also think it's important to call out the distinctive nature of accessibility in the context of Gutenberg. I like how you reflect on describing this as software accommodating users with "distinct characteristics and preferences for how to use said software". An example of this is reflected in the numerous entry points people have to insert a block. Is there some specific concern here around the use of the word "accessible" itself here?

File: docs/explanations/principles.md (hunk 14 above)
Comment (nerrad) [reply docs/explanations/principles.md]: Content wise in the description of the principle, or reflected in the principle statement itself?

File: docs/explanations/principles.md (hunk 15 above)
Comment (nerrad) [reply docs/explanations/principles.md:68]: So is this more a comment on the principle wording itself (that you would like changed) or on the description of the principle?

File: docs/README.md (hunk 7 above)
Comment (nerrad) [reply docs/README.md]: I'm going to go ahead and move this into the repo top-level README.md - I agree it makes more sense to target contributors. 

Given your [comment here however](https://github.com/WordPress/gutenberg/pull/30963#pullrequestreview-643571178), do you think I still need to expose this on on the built docs and if so, where do you think it should be in the docs structure?

File: docs/explanations/principles.md (hunk 1 above)
Comment (nerrad) [reply docs/explanations/principles.md]: See the changes made in d6db504.  I incorporated much of what Matías articulated in #18667 and this forms the basis for the revised content of the principle which is now simply "A canvas for expression" with the supporting points of this canvas:

- Guiding users around interacting with the content
- Uniformity within diversity
- and Accessbility

The wording around accessibility has also been tweaked in f8a44af which includes not only the work around this for _creators_ of content, but also for _consumers_ of content (to address what was [brought up here](https://github.com/WordPress/gutenberg/pull/30963#discussion_r620308866)).

This also handles the concerns around the word "delightful".

File: docs/explanations/principles.md (hunk 2 above)
Comment (nerrad) [reply docs/explanations/principles.md]: Addressed in 5e8d288.

File: docs/explanations/principles.md (hunk 3 above)
Comment (nerrad) [reply docs/explanations/principles.md:76]: Addressed in 5e8d288.

File: docs/explanations/principles.md (hunk 4 above)
Comment (nerrad) [reply docs/explanations/principles.md:78]: Addressed in 5e8d288.

File: docs/explanations/principles.md (hunk 5 above)
Comment (nerrad) [reply docs/explanations/principles.md]: Addressed in 5e8d288.

File: docs/README.md (hunk 7 above)
Comment (nerrad) [reply docs/README.md]: See 219f0af (and 1a648ba where wording was changed a bit).

File: docs/explanations/principles.md (hunk 8 above)
Comment (nerrad) [reply docs/explanations/principles.md:40]: See 5e744d9.

File: docs/explanations/principles.md (hunk 10 above)
Comment (nerrad) [reply docs/explanations/principles.md]: See 1a648ba.

File: docs/explanations/principles.md (hunk 11 above)
Comment (nerrad) [reply docs/explanations/principles.md]: See d6db504.

File: docs/explanations/principles.md (hunk 12 above)
Comment (nerrad) [reply docs/explanations/principles.md]: See d6db504.

File: docs/explanations/principles.md (hunk 13 above)
Comment (nerrad) [reply docs/explanations/principles.md:26]: See f8a44af.

File: docs/explanations/principles.md (hunk 14 above)
Comment (nerrad) [reply docs/explanations/principles.md]: This one was a bit difficult for me to address. I think in general I understand your feedback to mean we want to avoid the idea of complexity as a _starting point_ for the interactions. I've made a few changes in e7b3960 which I think helps, but it still feels like it could be better expressed.

File: docs/explanations/principles.md (hunk 15 above)
Comment (nerrad) [reply docs/explanations/principles.md:68]: I left this as is, it's not clear to me how to improve it.

**Comment by nerrad** [comment]
Sorry for the delay in addressing all the feedback. I've gone through and made changes and commented in the related threads to help highlight what was changed as a result. I think this is ready for another round of examination as folks are able.

File: docs/explanations/principles.md [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+
+Extensibility interfaces in Gutenberg favor:
+
+**Modularity** - the concept of being able to combine blocks or components into different types of layout or content (see “[Embrace the modularity](https://riad.blog/2020/01/28/embrace-the-modularity/)”).
+
+**Encapsulation** - there are appropriate error handling and boundaries around things that could be broken by things extending the interface.
+
+**Experimentation** - Often new APIs and interfaces clearly start as experimental while reliability and usability of different approaches are being tested. These experiments always begin with determining what problem is solved from the _creator_ perspective and seek to validate how it impacts the user workflow. The intent is to explore what generalizations of the creator's use-case are possible and to discover better abstractions before solidifying around the contract exposed to extension developers.
+
+For more information around currently built patterns for extensibility and key architectural concepts, see [this document](https://developer.wordpress.org/block-editor/explanations/architecture/key-concepts/)).
```
Comment (gziolo) [inline docs/explanations/principles.md]:
```suggestion
For more information around currently built patterns for extensibility and key architectural concepts, see [this document](/docs/explanations/architecture/key-concepts.md)).
```

File: docs/explanations/principles.md [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+
+Extensibility interfaces in Gutenberg favor:
+
+**Modularity** - the concept of being able to combine blocks or components into different types of layout or content (see “[Embrace the modularity](https://riad.blog/2020/01/28/embrace-the-modularity/)”).
+
+**Encapsulation** - there are appropriate error handling and boundaries around things that could be broken by things extending the interface.
+
+**Experimentation** - Often new APIs and interfaces clearly start as experimental while reliability and usability of different approaches are being tested. These experiments always begin with determining what problem is solved from the _creator_ perspective and seek to validate how it impacts the user workflow. The intent is to explore what generalizations of the creator's use-case are possible and to discover better abstractions before solidifying around the contract exposed to extension developers.
+
+For more information around currently built patterns for extensibility and key architectural concepts, see [this document](/docs/explanations/architecture/key-concepts.md)).
```
Comment (gziolo) [inline docs/explanations/principles.md]:
```suggestion
For more information around currently built patterns for extensibility and key architectural concepts, see [this document](/docs/explanations/architecture/key-concepts.md).
```

File: docs/explanations/principles.md (hunk 16 above)
Comment (nerrad) [reply docs/explanations/principles.md]: oops, thanks for making these changes to add the relative links @gziolo :)

File: docs/explanations/principles.md [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+
+Modern web application development brings its own set of accessibility problems around standards and consistent application between browsers which often requires some unique solutions that might not be apparent on the first attempt. So we acknowledge that accessibility is not something that can be implemented statically but instead is an ongoing part of the interface creation and refinement.
+
+## Curated Extensibility
+
+The use of the word _curated_ expressly refers to implementing extensibility in a way that prioritizes those _using Gutenberg interfaces to create their content_ over developers integrating with the system. As a result, every extensibility interface request is filtered through the question of how it impacts creator workflows.
```
Comment (getsource) [inline docs/explanations/principles.md]: I'm having a little bit of trouble parsing this paragraph.

In "_using Gutenberg interfaces to create their content_", does `interfaces` refer to UI/user interfaces, or interfaces as a whole, including APIs?

Assuming the prior, maybe "prioritizes those using Gutenberg user interfaces to create their content"?

Comment (nerrad) [reply docs/explanations/principles.md]: Good point. Fixed in 9b4eade (which also clarifies the usage of `interfaces` elsewhere in the doc).

Comment (getsource) [reply docs/explanations/principles.md]: Looks good; thanks much!

**Comment by nerrad** [comment]
Just checking in to see if there's any appetite for seeing this land from other @WordPress/gutenberg-core contributors. I let this sit for a bit given the efforts around getting WordPress 5.8 released. In particular, @mtias, I addressed some feedback you had and I'm not sure you've seen it, or if you have additional comments.

I'm happy to discuss/iterate more if needed (or even close the PR if there's no consensus).

**Comment by mtias** [comment]
Thanks for the nudge — I'm still not quite happy with some of the items and headings.

**Comment by nerrad** [comment]
> I'm still not quite happy with some of the items and headings.

Based on this, it's not clear to me what are the next steps. I've assumed don't merge. I also do not know which items and headings are still problematic or what needs done to make them better.

**Comment by mtias** [comment]
Sorry, I mean that some of the items in the outline feel at different levels but is not quite clear why. I don't have a concrete suggestion just yet. Perhaps it's trying to cover too much at once.

**Comment by gziolo** [comment]
That's very unfortunate to see this PR closed with no additions to the Block Editor Handbook 😞 

I was hoping we will finally have a good reference to use when discussing new extensibility points.

**Comment by getdave** [comment]
Agreed. How about after 5.9 we look to break this PR down into smaller documentation updates to various sections of the Block Editor Handbook? That way it will be easier to review and merge incrementally.

**Comment by nerrad** [comment]
Sorry folks, I closed this mostly because I felt it was stalled and I really had no clear idea on what some next steps could be taken to get it to a mergable state. While I do still think it'd be really helpful to have some master doc outlining principles guiding the project, I got a sense (maybe incorrectly) that there's still not sufficient consensus behind publishing this :). That's not necessarily a bad thing, it just means there's more work to do before something like this PR can be published. 

> How about after 5.9 we look to break this PR down into smaller documentation updates to various sections of the Block Editor Handbook? That way it will be easier to review and merge incrementally.

One thing I've noticed with the feedback that has come in, is that there _are_ some principles already sprinkled through various issues and docs already. Maybe all that is needed is to capture those and pull them into an index of sorts. The main outcome I was hoping from this PR was to have a single reference to point folks to for all the reasons I outlined in the PR description.

File: docs/explanations/principles.md [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+**Modularity** - the concept of being able to combine blocks or components into different types of layout or content (see “[Embrace the modularity](https://riad.blog/2020/01/28/embrace-the-modularity/)”).
+
+**Encapsulation** - there are appropriate error handling and boundaries around things that could be broken by things extending the interface.
+
+**Experimentation** - Often new APIs and interfaces clearly start as experimental while reliability and usability of different approaches are being tested. These experiments always begin with determining what problem is solved from the _creator_ perspective and seek to validate how it impacts the user workflow. The intent is to explore what generalizations of the creator's use-case are possible and to discover better abstractions before solidifying around the contract exposed to extension developers.
+
+For more information around currently built patterns for extensibility and key architectural concepts, see [this document](/docs/explanations/architecture/key-concepts.md).
+
+
+## Progressive Complexity unfolding from simple origins.
```
Comment (nerrad) [inline docs/explanations/principles.md:49]: Still not happy with this heading. Feels too wordy and abstract.

**Comment by mtias** [comment]
I personally don't mind this sitting open, we can come back to it, others can read it and add thoughts. Some time distance can also clarify what works and what doesn't.

**Comment by senadir** [comment]
Now that 5.9.0 is merged, would work on this get resumed?

**Comment by Mamaduka** [comment]
Should we close this PR, or is there a plan to continue working on core principles?

**Comment by priethor** [comment]
> Should we close this PR, or is there a plan to continue working on core principles?

It's been stale for more than 3 years, I think it's fine to close it and reopen it if we get to revisit the idea.
